File: packages/python/src/kymo/icons_pipeline.py
"""SVG normalize pipeline (P4 / CR-ICONS-005, FR-3/FR-6/FR-8).

`cleanup_svg → parse_colors(currentColor) → minify → validate`, plus
`to_record()` (→ sparse `{body, width, height}`) and `make_ids_safe()` for
`id`/`defs`-safe inlining (FR-7). Regex-based, stdlib-only — no SVGO/lxml
dependency (the JS mirror is `scripts/icons-pipeline.mjs`).

This is the machinery vectorization rides on; sourcing the vector originals
for the ~2,400 raster PNGs is the external prerequisite (RES-ICONS-001 §7.1),
so the catalogue migrates per set as originals land. `download --from iconify`
already runs `normalize()` on fetched art.
"""
from __future__ import annotations

import re

_DROP_BLOCKS = re.compile(
    r"<\?xml.*?\?>|<!DOCTYPE.*?>|<!--.*?-->|"
    r"<(script|style|metadata|title|desc)\b.*?</\1>",
    re.IGNORECASE | re.DOTALL,
)
_COLOR_ATTR = re.compile(r'\b(fill|stroke)\s*=\s*"(?!none|currentColor|url\()[^"]*"', re.IGNORECASE)
_WS = re.compile(r">\s+<")
_SVG_OPEN = re.compile(r"<svg\b[^>]*>", re.IGNORECASE | re.DOTALL)
_VIEWBOX = re.compile(r'viewBox\s*=\s*"([^"]+)"', re.IGNORECASE)
_WH = re.compile(r'\b(width|height)\s*=\s*"([0-9.]+)', re.IGNORECASE)


def cleanup_svg(text: str) -> str:
    """Strip XML decl / doctype / comments / scripts / styles / editor cruft."""
    return _DROP_BLOCKS.sub("", text).strip()


def parse_colors(text: str) -> str:
    """Recolour concrete fill/stroke to `currentColor` (themeable, FR-6).
    Leaves `none`, existing `currentColor`, and `url(...)` refs untouched."""
    return _COLOR_ATTR.sub(lambda m: f'{m.group(1)}="currentColor"', text)


def minify(text: str) -> str:
    """Collapse inter-tag whitespace."""
    return _WS.sub("><", text).strip()


def normalize(text: str) -> str:
    """Full pipeline returning a cleaned, recoloured SVG document."""
    return minify(parse_colors(cleanup_svg(text)))


def _dims(open_tag: str) -> tuple[int, int]:
    vb = _VIEWBOX.search(open_tag)
    if vb:
        parts = vb.group(1).replace(",", " ").split()
        if len(parts) == 4:
            return round(float(parts[2])), round(float(parts[3]))
    wh = dict((m.group(1).lower(), float(m.group(2))) for m in _WH.finditer(open_tag))
    return round(wh.get("width", 24)), round(wh.get("height", 24))


def to_record(text: str) -> dict:
    """Normalize, then return the sparse IconifyJSON record `{body, width,
    height}` — the inner body only, no `<svg>` wrapper (FR-3)."""
    doc = normalize(text)
    m = _SVG_OPEN.search(doc)
    if m:
        body = doc[m.end():]
        body = re.sub(r"</svg>\s*$", "", body, flags=re.IGNORECASE).strip()
        width, height = _dims(m.group(0))
    else:                              # already a bare body
        body, (width, height) = doc, (24, 24)
    return {"body": body, "width": width, "height": height}


_ID_DEF = re.compile(r'\bid\s*=\s*"([^"]+)"')


def make_ids_safe(body: str, suffix: str) -> str:
    """Suffix every element `id` (and its `url(#id)` / `href="#id"` refs) so
    the same icon inlined N times in one document never collides (FR-7)."""
    ids = set(_ID_DEF.findall(body))
    for old in ids:
        new = f"{old}-{suffix}"
        body = re.sub(rf'(\bid\s*=\s*")({re.escape(old)})(")', rf"\g<1>{new}\g<3>", body)
        body = body.replace(f"url(#{old})", f"url(#{new})")
        body = re.sub(rf'((?:xlink:)?href\s*=\s*")#{re.escape(old)}(")', rf"\g<1>#{new}\g<2>", body)
    return body

File: packages/js/scripts/icons-pipeline.mjs
/**
 * SVG normalize pipeline — JS mirror of packages/python/src/kymo/icons_pipeline.py
 * (P4 / CR-ICONS-005, FR-3/FR-6/FR-7/FR-8). Regex-based, ZERO dependencies.
 *
 *   cleanupSvg → parseColors(currentColor) → minify  (+ toRecord, makeIdsSafe)
 *
 * Vectorization rides on this; sourcing vector originals for the raster PNGs
 * is the external prerequisite (RES-ICONS-001 §7.1). `kymo icons download
 * --from iconify` runs normalize() on fetched art.
 */
const DROP_BLOCKS = /<\?xml[\s\S]*?\?>|<!DOCTYPE[\s\S]*?>|<!--[\s\S]*?-->|<(script|style|metadata|title|desc)\b[\s\S]*?<\/\1>/gi;
const COLOR_ATTR = /\b(fill|stroke)\s*=\s*"(?!none|currentColor|url\()[^"]*"/gi;
const WS = />\s+</g;
const SVG_OPEN = /<svg\b[^>]*>/i;
const VIEWBOX = /viewBox\s*=\s*"([^"]+)"/i;
const WH = /\b(width|height)\s*=\s*"([0-9.]+)/gi;
const ID_DEF = /\bid\s*=\s*"([^"]+)"/g;

export const cleanupSvg = (t) => t.replace(DROP_BLOCKS, "").trim();
export const parseColors = (t) => t.replace(COLOR_ATTR, (_m, a) => `${a}="currentColor"`);
export const minify = (t) => t.replace(WS, "><").trim();
export const normalize = (t) => minify(parseColors(cleanupSvg(t)));

function dims(openTag) {
  const vb = VIEWBOX.exec(openTag);
  if (vb) {
    const p = vb[1].replace(/,/g, " ").trim().split(/\s+/);
    if (p.length === 4) return [Math.round(+p[2]), Math.round(+p[3])];
  }
  const wh = {};
  for (const m of openTag.matchAll(WH)) wh[m[1].toLowerCase()] = +m[2];
  return [Math.round(wh.width ?? 24), Math.round(wh.height ?? 24)];
}

export function toRecord(text) {
  const doc = normalize(text);
  const m = SVG_OPEN.exec(doc);
  if (!m) return { body: doc, width: 24, height: 24 };
  const body = doc.slice(m.index + m[0].length).replace(/<\/svg>\s*$/i, "").trim();
  const [width, height] = dims(m[0]);
  return { body, width, height };
}

export function makeIdsSafe(body, suffix) {
  const ids = new Set([...body.matchAll(ID_DEF)].map((m) => m[1]));
  let out = body;
  for (const old of ids) {
    const safe = old.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const next = `${old}-${suffix}`;
    out = out.replace(new RegExp(`(\\bid\\s*=\\s*")(${safe})(")`, "g"), `$1${next}$3`);
    out = out.split(`url(#${old})`).join(`url(#${next})`);
    out = out.replace(new RegExp(`((?:xlink:)?href\\s*=\\s*")#${safe}(")`, "g"), `$1#${next}$2`);
  }
  return out;
}
